File: apps/backend/internal/models/graph.go
package models

import "time"

// GraphResponse contains graph data for visualization
type GraphResponse struct {
	Nodes []GraphNode `json:"nodes"`
	Edges []GraphEdge `json:"edges"`
	Chain ChainInfo   `json:"chain"`
}

// GraphNode represents a principal in the graph
type GraphNode struct {
	ID       string                 `json:"id"`
	Label    string                 `json:"label"`
	Type     string                 `json:"type"` // "root", "intermediate", "leaf"
	Level    int                    `json:"level"`
	Metadata map[string]interface{} `json:"metadata"`
}

// GraphEdge represents a delegation relationship
type GraphEdge struct {
	Source     string                 `json:"source"`
	Target     string                 `json:"target"`
	Capability CapabilityInfo         `json:"capability"`
	Valid      bool                   `json:"valid"`
	Label      string                 `json:"label"`
	Level      int                    `json:"level"`
	Type       string                 `json:"type"` // delegation, invocation, proof
	Metadata   map[string]interface{} `json:"metadata,omitempty"`
}
// InvocationGraphResponse contains graph data for invocation visualization
type InvocationGraphResponse struct {
	Nodes        []GraphNode          `json:"nodes"`
	Edges        []GraphEdge          `json:"edges"`
	Chain        ChainInfo            `json:"chain"`
	Invocation   *InvocationResponse  `json:"invocation"`
	IsInvocation bool                 `json:"isInvocation"`
}



// PrincipalInfo represents a DID in the chain
type PrincipalInfo struct {
	DID   string   `json:"did"`
	Role  string   `json:"role"` // "root", "intermediate", "leaf"
	Level int      `json:"level"`
	CIDs  []string `json:"cids"`
}

// TimelineEvent represents a temporal event in the delegation chain
type TimelineEvent struct {
	Type      string    `json:"type"`      // "issued", "expires"
	Time      time.Time `json:"time"`
	CID       string    `json:"cid"`
	Level     int       `json:"level"`
	Principal string    `json:"principal"`
}

File: apps/backend-js/src/services/graph.js
export class GraphService {
  constructor(parserService) {
    this.parser = parserService;
  }

  async generateDelegationGraph(tokenBytes) {
    const chain = await this.parser.parseDelegationChain(tokenBytes);
    const { nodes, edges } = this.#buildGraph(chain);
    const chainInfo = this.#buildChainInfo(chain);

    return { nodes, edges, chain: chainInfo };
  }

  async generateInvocationGraph(tokenBytes) {
    const invocation = await this.parser.parseInvocation(tokenBytes);
    const chain = await this.parser.parseDelegationChain(tokenBytes);
    const { nodes, edges } = this.#buildGraph(chain);
    const chainInfo = this.#buildChainInfo(chain);

    return {
      nodes,
      edges,
      chain: chainInfo,
      invocation,
      isInvocation: invocation.isInvocation
    };
  }

  #buildGraph(chain) {
    const nodesMap = new Map();
    const edges = [];
    const maxLevel = Math.max(...chain.map(d => d.level));

    for (const del of chain) {
      if (!nodesMap.has(del.issuer)) {
        nodesMap.set(del.issuer, {
          id: del.issuer,
          label: this.#shortenDID(del.issuer),
          type: del.level === 0 ? 'root' : 'intermediate',
          level: del.level,
          metadata: { fullDID: del.issuer, role: 'delegator' }
        });
      }

      if (!nodesMap.has(del.audience)) {
        nodesMap.set(del.audience, {
          id: del.audience,
          label: this.#shortenDID(del.audience),
          type: del.level === maxLevel ? 'leaf' : 'intermediate',
          level: del.level,
          metadata: { fullDID: del.audience, role: 'delegatee' }
        });
      }

      for (const cap of del.capabilities) {
        edges.push({
          source: del.issuer,
          target: del.audience,
          capability: cap,
          label: `${cap.can} on ${cap.with}`,
          valid: true,
          level: del.level,
          type: 'delegation',
          metadata: { cid: del.cid }
        });
      }
    }

    return {
      nodes: Array.from(nodesMap.values()),
      edges
    };
  }

  #buildChainInfo(chain) {
    if (chain.length === 0) return {};

    const maxLevel = Math.max(...chain.map(d => d.level));
    const leafCIDs = chain.filter(d => d.level === maxLevel).map(d => d.cid);

    return {
      totalLevels: maxLevel + 1,
      isComplete: true,
      rootCID: chain[0].cid,
      leafCIDs,
      principals: [],
      timeline: []
    };
  }

  #shortenDID(did) {
    if (did.length <= 20) return did;
    const parts = did.split(':');
    if (parts.length < 3) return did.slice(0, 20) + '...';
    
    const identifier = parts[2];
    if (identifier.length > 16) {
      return `did:${parts[1]}:${identifier.slice(0, 8)}...${identifier.slice(-8)}`;
    }
    return did;
  }
}
